microservice/package/src/main/java/com/cumulocity/agent/packaging/microservice/impl/MicroserviceDockerClientImpl.java

```diff
@@ -1,81 +1,261 @@
-    private DockerClient docker;
 
-    public void saveDockerImage(final String image, final File targetFile) throws InterruptedException, DockerException, IOException {
-        try (
-                final InputStream inputStream = docker.save(image);
-                final FileOutputStream outputStream = new FileOutputStream(targetFile)
-        ) {
+    DockerClient dockerClient;
+
+    @SneakyThrows
+    public void buildDockerImage(String dockerDirectory, Set<String> tags, Map<String, String> buildArgs, String networkMode, Integer dockerBuildTimeout) {
+
+        BuildImageCmd buildImageCmd = dockerClient.buildImageCmd(new File(dockerDirectory)).withTags(tags);
+
+        for (Map.Entry<String, String> buildArgument : buildArgs.entrySet()) {
+            buildImageCmd = buildImageCmd.withBuildArg(buildArgument.getKey(), buildArgument.getValue());
+        }
+
+        if (!Strings.isNullOrEmpty(networkMode)) {
+            buildImageCmd = buildImageCmd.withNetworkMode(networkMode);
+        }
+
+        log.info("Building Docker image. Docker dir={},tags={}, build arguments={}", dockerDirectory, tags, buildArgs);
+        ImageBuildCompletionWaiter imageBuildCompletionWaiter = new ImageBuildCompletionWaiter();
+        buildImageCmd.exec(imageBuildCompletionWaiter);
+        log.info("Waiting for image build to complete (timeout={}s)", dockerBuildTimeout);
+        imageBuildCompletionWaiter.awaitWithTimeout(dockerBuildTimeout);
+
+        log.info("Image build success");
+
+
+    }
+
+    protected class ImageBuildCompletionWaiter extends BuildImageResultCallback {
+
+        String completedImageId;
+        boolean buildFailed;
+        private Future f;
+
+        public void awaitWithTimeout(int seconds) throws MavenExecutionException {
+
+            ExecutorService executorService = Executors.newSingleThreadExecutor();
+            f = executorService.submit(() -> {
+
+                while (Objects.isNull(completedImageId)) {
+                    try {
+                        Thread.sleep(250);
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                }
+
+                log.info("Build successful, imageId: {}", completedImageId);
+            });
+
+            try {
+                f.get(seconds, TimeUnit.SECONDS);
+            } catch (TimeoutException timeoutException) {
+                throw new MavenExecutionException("Docker build timed out", timeoutException);
+            } catch (ExecutionException e) {
+                throw new MavenExecutionException("Docker build execution failed", e);
+            } catch (CancellationException c) {
+                throw new MavenExecutionException("Docker image build failed. See logs above", c);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```

Comment:
the InterruptedException exception should not be ignored. 
You must call : 
Thread.currentThread().interrupt();
And you may propagete the ExecutionException or RuntimeException 

Otherwise if someone would like to interrupt the process with CTRL-C it maybe not stop